cocos2d/Platforms/Mac/CCViewMacGL.m

```diff
@@ -0,0 +1,296 @@
+    // When resizing the view, -reshape is called automatically on the main thread
+    // Add a mutex around to avoid the threads accessing the context simultaneously when resizing
+    
+    [self lockOpenGLContext];
+    
+    NSRect rect = [self convertRectToBacking:self.bounds];
+    
+    CCDirector *director = _director;
+    [director reshapeProjection: NSSizeToCGSize(rect.size) ];
+    
+    // avoid flicker
+    // Only draw if there is something to draw, otherwise it actually creates a flicker of the current glClearColor
+    if(director.runningScene){
+        [director mainLoopBody];
+    }
+    //	[self setNeedsDisplay:YES];
+    
+    [self unlockOpenGLContext];
+}
+
+
+-(void) lockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    [glContext makeCurrentContext];
+    CGLLockContext([glContext CGLContextObj]);
+}
+
+-(void) unlockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    CGLUnlockContext([glContext CGLContextObj]);
+}
+
+// Find or make a fence that is ready to use.
+-(CCGLFence *)getReadyFence
+{
+    // First checkf oldest (first in the array) fence is ready again.
+    CCGLFence *fence = _fences.firstObject;;
+    if(fence.isReady){
+        // Remove the fence so it can be inserted at the end of the queue again.
+        [_fences removeObjectAtIndex:0];
+        return fence;
+    } else {
+        // No existing fences ready. Make a new one.
+        return [[CCGLFence alloc] init];
+    }
+}
+
+-(void)addFrameCompletionHandler:(dispatch_block_t)handler
+{
+    if(_fences == nil){
+        _fences = [NSMutableArray arrayWithObject:[[CCGLFence alloc] init]];
+    }
+    
+    CCGLFence *fence = _fences.lastObject;
+    if(!fence.isReady){
+        fence = [self getReadyFence];
+        [_fences addObject:fence];
+    }
+    
+    [fence.handlers addObject:handler];
+}
+
+-(void)beginFrame
+{
+    [self lockOpenGLContext];
+}
+
+-(void)presentFrame
+{
+    {
+        CCGLFence *fence = _fences.lastObject;
+        if(fence.isReady){
+            // If the fence is ready to be added, insert a sync point for it.
+            [fence insertFence];
+        }
+    }
+    
+    [self.openGLContext flushBuffer];
+    
+    // Check the fences for completion.
+    for(CCGLFence *fence in _fences){
+        if(fence.isCompleted){
+            for(dispatch_block_t handler in fence.handlers) handler();
+            [fence.handlers removeAllObjects];
+        } else {
+            break;
+        }
+    }
+    
+    [self unlockOpenGLContext];
+}
+
+-(GLuint)fbo
+{
+    return 0;
+}
+
+- (void) dealloc
+{
+    CCLOGINFO(@"cocos2d: deallocing %@", self);
+}
+
+#pragma mark CCGLView - Mouse Delegate
+
+- (void)mouseDown:(NSEvent *)theEvent
+{
+    // dispatch mouse to responder manager
+    [CCDirector bindDirector:_director];
```

Comment:
Good ideas! The extra mirrored methods in CCScheduler are definitely superfluous. I'll update the request. The CCViewMacGL has way more events than iOS, they can definitely be sent to a much smaller number of already existing methods. It could be converted to mouse, scroll wheel, and key events pretty trivially, and could be convered to a single handleUIEvent method too (although organizationally that might be a bit of a big method).

Yeah, I should have removed those nonsense comments when I was touching it. CCViewMacGL is a renamed CCGLView, which is pretty old. Will cleanup.

There are situations where the same thread could have different directors (and different views). This is potentially addressing situations where a game is running with two CCViews, each one would have different directors. I'm guessing we can't make any assumptions over which thread is calling these input methods. Maybe @slembcke can comment more on this. It also improves testability, when we can create fresh CCDirectors for each test.  Any other ideas for improving this? The only one that really comes to mind is "bring back the singleton" or do a better job of encapsulating the bindDirector calls (which I am on!).